### ISSUE TITLE
Memory leak when file appender can't write

### ISSUE THREAD

**Issue by jsha** [log4js-node/log4js-node#620]
When a file appender can't write to its file, log4js leaks memory.

Steps to reproduce:

```
$ sudo touch cheese.log
$ ls -l cheese.log
-rw-r--r-- 1 root root 0 Nov 24 22:31 cheese.log
$ node test/memory-leak.js &
$ while ps -o rss,command | grep "node[ ]test/memory-leak" ; do sleep 60 ; done 
68440 node test/memory-leak.js
83408 node test/memory-leak.js
114292 node test/memory-leak.js
128812 node test/memory-leak.js
143332 node test/memory-leak.js
```

I mentioned at https://github.com/log4js-node/log4js-node/issues/575#issuecomment-346921386 that I though this might be due to `console.error` triggering a memory leak in node, but after some more testing, I see that `console.error` seems to get called just once.

By comparison, if cheese.log is writeable, there is no memory leak.

memory-leak.js:
```
"use strict";
const log4js = require('../lib/log4js');
log4js.configure({
  appenders: [{ type: 'file', filename: 'cheese.log', category: 'cheese' }]
});
const logger = log4js.getLogger('cheese');
setInterval(() => logger.info('aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa'), 1);
```

**Comment by jsha** [comment]
After looking at some heap dumps, this looks like we are retaining a large buffer in `stream.Writeable`, which is used in log4js by way of `BaseRollingFileStream`. You can reproduce the same memory leak with:

```
const fs = require('fs');
const stream = fs.createWriteStream('cheese.log');
stream.on("error", console.error);
setInterval(() => stream.write('aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa'), 1)
```

In the heap dumps, it looks like the stream is "corked", and all the data written to it is being buffered in anticipation of an uncorking that will never come.

It appears the way this is dealt with in `stream` is by returning a boolean from `stream.write`:

https://nodejs.org/api/stream.html#stream_writable_write_chunk_encoding_callback
> Returns: <boolean> false if the stream wishes for the calling code to wait for the 'drain' event to be emitted before continuing to write additional data; otherwise true.

I think the fix for this in log4js should be to pay attention to the return value from `stream.write`, and throw away logs until it receives a `drain` event. If the user wishes to hang onto more than the default amount of logs in hopes that an error condition is recoverable, the stream's `highWaterMark` can be configured (though it's probably not worth the complexity to plumb it through).

**Comment by nomiddlename** [comment]
Thanks for catching this (and the PR), my only worry is the idea of having to discard log messages when they can't be written. But I guess that losing log messages is better than having your app crash. I should also tie this in to how we handle errors coming from the appenders - #575

**Comment by nomiddlename** [comment]
This will be fixed by PR #938 - allowing the user to listen for "log4js:pause" events and deciding if they want to pause logging or not.

**Comment by nomiddlename** [comment]
The "log4js:pause" event changes to file and dateFile appenders was published to npm in log4js@6.1.0 - take a look and see if it helps your situation.

**Comment by nomiddlename** [comment]
Going to close this one, please open a new issue if the "log4js:pause" event does not help.
